### PR TITLE
feat(ios): display monthly balance and rollover separately in budget lists

### PR DESCRIPTION
**Pull request by neogenz**
## Summary

- Add `monthlyBalance` computed property to separate pure monthly result from rollover
- Update `BudgetListStore` sparse fields to fetch `rollover` for dashboard display
- Display rollover as separate badge in `BudgetMonthRow` and `CurrentMonthHeroCard`
- Add comprehensive unit tests for `BudgetSparse.monthlyBalance` edge cases

## Type

feat

### CODE REVIEW

File: ios/Pulpe/Features/Budgets/BudgetDetails/RolloverInfoRow.swift [hunk 1]
```diff
@@ -6,50 +6,104 @@ struct RolloverInfoRow: View {
...
 
+    private var accentColor: Color {
+        isPositive ? .financialSavings : .financialOverBudget
+    }
+
+    private var label: String {
+        isPositive ? "Excédent reporté" : "Déficit reporté"
+    }
+
+    @Environment(\.colorScheme) private var colorScheme
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/Budgets/BudgetDetails/RolloverInfoRow.swift]: Fixed in cd17c607 — removed the unused `@Environment(\.colorScheme)`.

File: ios/Pulpe/Features/Budgets/BudgetDetails/BudgetDetailsView.swift [hunk 2]
```diff
@@ -178,31 +178,22 @@ struct BudgetDetailsView: View {
...
+            // Hero balance card (with integrated rollover)
             Section {
                 HeroBalanceCard(
                     metrics: viewModel.metrics,
-                    timeElapsedPercentage: timeElapsedPercentage
+                    timeElapsedPercentage: timeElapsedPercentage,
+                    rolloverAmount: viewModel.rolloverInfo?.amount,
+                    onRolloverTap: viewModel.rolloverInfo?.previousBudgetId.map { id in
+                        { previousBudgetItem = PreviousBudgetItem(id: id) }
+                    } ?? nil
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/Budgets/BudgetDetails/BudgetDetailsView.swift]: Already fixed in 5235c3a0 — the `?? nil` was removed in that earlier commit.

File: ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift [hunk 3]
```diff
@@ -5,6 +5,8 @@ struct HeroBalanceCard: View {
     let metrics: BudgetFormulas.Metrics
     var timeElapsedPercentage: Double = 0
     var onTapProgress: (() -> Void)?
+    var rolloverAmount: Decimal?
+    var onRolloverTap: (() -> Void)?
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift:10]: Fixed in cd17c607 — added `accessibilityAction(named: "Voir le budget précédent")` via `.ifLet(onRolloverTap)`.

File: ios/Pulpe/Features/Budgets/BudgetDetails/BudgetDetailsView.swift [hunk 4]
```diff
@@ -129,6 +130,12 @@ struct BudgetDetailsView: View {
         .sheet(item: $previousBudgetItem) { item in
             PreviousBudgetSheet(budgetId: item.id)
         }
+        .sheet(isPresented: $showRealizedBalance) {
+            RealizedBalanceSheet(
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/Budgets/BudgetDetails/BudgetDetailsView.swift:134]: Fixed in fd4a2f7d — moved `RealizedBalanceSheet` from `Features/CurrentMonth/Components/` to `Shared/Components/`.

File: ios/Pulpe/Shared/Components/CurrencyField.swift [hunk 5]
```diff
@@ -62,15 +62,15 @@ struct CurrencyField: View {
...
                 TextField(hint, text: $textValue)
                     .keyboardType(.decimalPad)
                     .foregroundStyle(Color.authInputText)
                     .focused(externalFocus ?? $internalFocus)
                     .accessibilityLabel(label ?? "Montant en CHF")
+
+                Text("CHF")
+                    .foregroundStyle(prefixColor)
+                    .font(PulpeTypography.bodyLarge)
                     .onChange(of: textValue) { _, newValue in
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Shared/Components/CurrencyField.swift:70]: Fixed in fd4a2f7d — moved `.onChange` modifiers from `Text("CHF")` to `TextField`.

File: ios/Pulpe/Domain/Models/Budget.swift [hunk 6]
```diff
@@ -127,6 +127,12 @@ struct BudgetSparse: Decodable, Identifiable, Sendable, Hashable {
                year == calendar.component(.year, from: now)
     }
 
+    /// Pure monthly result: income - expenses (incl. savings) without rollover
+    var monthlyBalance: Decimal? {
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Domain/Models/Budget.swift]: Acknowledged — `monthlyBalance` is intentionally kept: it's well-tested (7 tests) and prepares for the monthly balance report feature. No production consumer yet, but the property is stable and ready to plug in.

File: ios/Pulpe/Features/Budgets/BudgetDetails/BudgetDetailsViewModel.swift [hunk 7]
```diff
@@ -181,6 +187,25 @@ final class BudgetDetailsViewModel {
         }
     }
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/Budgets/BudgetDetails/BudgetDetailsViewModel.swift:189]: Fixed in fd4a2f7d — marked `displayBudgetLines` as `private` in `BudgetDetailsViewModel`.

File: ios/PulpeTests/Domain/Formulas/BudgetFormulasRealizedTests.swift [hunk 8]
```diff
@@ -30,6 +30,59 @@ struct BudgetFormulasRealizedTests {
...
+        let balance = BudgetFormulas.calculateRealizedBalance(budgetLines: lines)
+        #expect(balance == 50)
+    }
+
+    @Test func calculateRealizedBalance_includesCheckedPositiveRollover() {
+        let lines = [
+            TestDataFactory.createBudgetLine(id: "1", amount: 5000, kind: .income, isChecked: true),
+            TestDataFactory.createBudgetLine(id: "2", amount: 3094, kind: .income, isChecked: true, isRollover: true),
+            TestDataFactory.createBudgetLine(id: "3", amount: 8000, kind: .expense, isChecked: true)
+        ]
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/PulpeTests/Domain/Formulas/BudgetFormulasRealizedTests.swift:60]: Fixed in fd4a2f7d — factory now uses `abs(amount)` so a -1950 rollover produces `amount: 1950, kind: .expense`. Test updated to use `BudgetLine.rolloverLine(amount: -1950, ...)` instead of manual creation.

File: ios/Pulpe/Features/Budgets/BudgetDetails/BudgetDetailsViewModel.swift [hunk 9]
```diff
@@ -181,6 +187,25 @@ final class BudgetDetailsViewModel {
...
+        guard let budget, let rollover = budget.rollover, rollover != 0 else {
+            return budgetLines
+        }
+        let rolloverLine = BudgetLine.rolloverLine(
+            amount: rollover,
+            budgetId: budget.id,
+            sourceBudgetId: budget.previousBudgetId
+        )
+        return [rolloverLine] + budgetLines
+    }
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/Budgets/BudgetDetails/BudgetDetailsViewModel.swift:192]: Fixed in fd4a2f7d — extracted shared logic into `Budget.displayBudgetLines(base:)`. Both `CurrentMonthStore` and `BudgetDetailsViewModel` now delegate to it.

File: ios/Pulpe/Domain/Models/Budget.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Domain/Models/Budget.swift]: Fixed in 6c5dac87 — removed `monthlyBalance` and its test suite (YAGNI: no production consumer).

File: ios/Pulpe/Domain/Models/Budget.swift [hunk 10]
```diff
@@ -49,6 +49,18 @@ struct Budget: Codable, Identifiable, Hashable, Sendable {
...
     }
+
+    /// Budget lines augmented with a virtual rollover line when rollover != 0.
+    /// Shared logic used by both CurrentMonthStore and BudgetDetailsViewModel.
+    func displayBudgetLines(base: [BudgetLine]) -> [BudgetLine] {
+        guard let rollover, rollover != 0 else { return base }
+        let rolloverLine = BudgetLine.rolloverLine(
+            amount: rollover,
+            budgetId: id,
+            sourceBudgetId: previousBudgetId
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Domain/Models/Budget.swift]: Fixed in 6c5dac87 — moved `displayBudgetLines` to `BudgetFormulas` as a static function. Both `CurrentMonthStore` and `BudgetDetailsViewModel` now call `BudgetFormulas.displayBudgetLines(base:budget:)`.

File: ios/Pulpe/Domain/Store/CurrentMonthStore.swift [hunk 11]
```diff
@@ -507,16 +507,7 @@ extension CurrentMonthStore {
...
-            return budgetLines
-        }
-        // Add virtual rollover line
-        let rolloverLine = BudgetLine.rolloverLine(
-            amount: rollover,
-            budgetId: budget.id,
-            sourceBudgetId: budget.previousBudgetId
-        )
-        return [rolloverLine] + budgetLines
+        budget?.displayBudgetLines(base: budgetLines) ?? budgetLines
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Domain/Store/CurrentMonthStore.swift]: Fixed in 6c5dac87 — the method now lives in `BudgetFormulas.displayBudgetLines(base:budget:)`. Both call sites use the static function, build passes.

File: ios/Pulpe/Features/Budgets/BudgetDetails/BudgetDetailsViewModel.swift [hunk 12]
```diff
@@ -181,6 +187,17 @@ final class BudgetDetailsViewModel {
         }
     }
 
+    private var displayBudgetLines: [BudgetLine] {
+        budget?.displayBudgetLines(base: budgetLines) ?? budgetLines
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/Budgets/BudgetDetails/BudgetDetailsViewModel.swift]: Fixed in 6c5dac87 — same fix as above, `BudgetFormulas.displayBudgetLines(base:budget:)` resolves both call sites.

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -615,6 +615,12 @@ jobs:
           elif grep -q '\*\* TEST FAILED \*\*' test-output.log; then
             echo "❌ Tests failed"
             exit 1
+          elif grep -q '✘' test-output.log; then
+            echo "❌ Test failures detected (✘ found in output)"
+            exit 1
+          elif grep -q '✔' test-output.log && ! grep -q '✘' test-output.log; then
```
[the remark this answers is not in the document]
Comment (neogenz) [reply .github/workflows/ci.yml:621]: Fixed in 5b80ebfe — added PASS_COUNT guard (threshold ≥ 50) to detect partial test runs. Uses `grep -c '✔'` with fallback to 0.

File: ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift [hunk 14]
```diff
@@ -170,6 +206,66 @@ struct HeroBalanceCard: View {
...
+    // MARK: - Rollover Footer
+
+    private var rolloverIcon: String {
+        guard let rolloverAmount else { return "arrow.triangle.2.circlepath" }
+        if rolloverAmount > 0 { return "arrow.up.right.circle" }
+        if rolloverAmount < 0 { return "arrow.down.right.circle" }
+        return "equal.circle"
+    }
+
+    private func rolloverFooter(amount: Decimal) -> some View {
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift:211]: Fixed in 5b80ebfe — added `@ViewBuilder`, removed `Group` wrapper, moved `.padding` inside each branch (semantically identical since Group is layout-transparent).

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -615,6 +615,17 @@ jobs:
...
+            PASS_COUNT=$(grep -c '✔' test-output.log || echo 0)
+            echo "⚠️ xcodebuild hung before writing summary — ${PASS_COUNT} tests passed, no failures detected"
+            if [ "$PASS_COUNT" -lt 50 ]; then
```
[the remark this answers is not in the document]
Comment (neogenz) [reply .github/workflows/ci.yml]: Fixed in 27e21062 — added inline comment: `# ~120 tests currently — 50 catches partial runs`

File: ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift [hunk 16]
```diff
@@ -86,15 +89,34 @@ struct HeroBalanceCard: View {
...
             return "\(contextLabel) — montant masqué. \(motivationalMessage)"
         }
-        return """
+        var desc = """
         \(contextLabel) \(formattedBalance) CHF. \
         \(motivationalMessage). \
         Dépensé \(formattedSpent) sur \(formattedAvailable)
         """
+        if let rolloverAmount {
+            desc += ". \(rolloverLabel) de \(rolloverAmount.asCHF)"
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift]: Fixed in dc9d4f14 — wrapped `rolloverAmount` with `abs()` in the accessibility description to avoid VoiceOver reading a redundant negative sign.

File: ios/Pulpe/Features/Budgets/BudgetDetails/RolloverInfoRow.swift [hunk 17]
```diff
@@ -6,50 +6,102 @@ struct RolloverInfoRow: View {
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/Budgets/BudgetDetails/RolloverInfoRow.swift]: Already resolved — `RolloverInfoRow.swift` was deleted in 0a1e82b7 (no remaining call sites in the codebase).

File: ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift [hunk 18]
```diff
@@ -86,15 +89,34 @@ struct HeroBalanceCard: View {
         }
     }
 
+    /// Tint color for glass overlays — matches the hero gradient per emotion state
+    /// so Liquid Glass blends into the card instead of appearing white.
+    private var glassTintColor: Color {
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift:94]: Already resolved — `glassTintColor` uses named design tokens (`.heroTintComfortable`, `.heroTintTight`, `.heroTintDeficit`) instead of hard-coded hex values.

File: ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift [hunk 19]
```diff
@@ -170,6 +206,64 @@ struct HeroBalanceCard: View {
         .animation(.spring(response: 0.7, dampingFraction: 0.8), value: metrics.emotionState)
     }
 
+    // MARK: - Rollover Footer
+
+    private var rolloverIcon: String {
+        guard let rolloverAmount else { return "arrow.triangle.2.circlepath" }
+        return rolloverAmount >= 0 ? "arrow.up.right.circle" : "arrow.down.right.circle"
+    }
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift]: Fixed in dc9d4f14 — inlined the icon ternary directly in `rolloverPill`, removed the standalone `rolloverIcon(for:)` function.

File: .github/workflows/ci.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply .github/workflows/ci.yml]: Fixed in 27e21062 — added named variable `EXPECTED_MIN_TESTS=50` with a comment documenting the rationale.

File: ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift [hunk 20]
```diff
@@ -117,6 +135,9 @@ struct HeroBalanceCard: View {
         .accessibilityElement(children: .ignore)
         .accessibilityLabel(accessibilityDescription)
         .accessibilityAddTraits(onTapProgress != nil ? .isButton : [])
+        .ifLet(onRolloverTap) { view, action in
+            view.accessibilityAction(named: "Voir le budget précédent", action)
+        }
```
[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift:140]: Fixed in ff4ab97c — added `.accessibilityAction(named: "Suivi du budget")` so VoiceOver exposes the chart action despite `.accessibilityElement(children: .ignore)`.

[the remark this answers is not in the document]
Comment (neogenz) [reply ios/Pulpe/Features/CurrentMonth/Components/HeroBalanceCard.swift:140]: Duplicate of above — fixed in ff4ab97c.
